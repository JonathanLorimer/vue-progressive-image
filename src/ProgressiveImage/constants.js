export const PLACEHOLDER_LOAD_SUCCESS = "placeholder-load";
export const PLACEHOLDER_LOAD_FAIL = "placeholder-fail";
export const MAIN_IMAGE_LOAD_SUCCESS = "load";
export const PLUGIN_PREFIX = "v-progressive-image";

export const DEFAULT_IMAGE_POLL_INTERVAL = 10;
export const DEFAULT_IMAGE_BLUR = 20;
export const DEFAULT_IMAGE_RENDERING_DELAY = 0;
export const DEFAULT_IMAGE_ASPECT_RATIO = 0.5625;
