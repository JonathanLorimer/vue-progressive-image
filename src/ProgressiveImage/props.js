import {
  DEFAULT_IMAGE_RENDERING_DELAY,
  DEFAULT_IMAGE_BLUR,
  DEFAULT_IMAGE_ASPECT_RATIO,
  DEFAULT_IMAGE_POLL_INTERVAL,
} from "./constants";

export default {
  src: {
    type: String,
    required: true,
  },

  placeholderSrc: {
    type: String,
  },

  fallbackSrc: {
    type: String,
  },

  delay: {
    type: [Number, String],
    default: DEFAULT_IMAGE_RENDERING_DELAY,
  },

  blur: {
    type: [Number, String],
    default: DEFAULT_IMAGE_BLUR,
  },

  alt: {
    type: String,
  },

  circle: {
    type: Boolean,
    default: false,
  },

  objectCover: {
    type: Boolean,
    default: false,
  },

  objectContain: {
    type: Boolean,
    default: false,
  },

  selectNone: {
    type: Boolean,
    default: false,
  },

  aspectRatio: {
    type: [Number, String],
    default: DEFAULT_IMAGE_ASPECT_RATIO,
  },

  pollInterval: {
    type: [Number, String],
    default: DEFAULT_IMAGE_POLL_INTERVAL,
  },
};
