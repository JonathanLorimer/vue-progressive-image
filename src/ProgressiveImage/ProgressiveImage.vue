<script setup>
import { ref, onMounted, computed } from "vue";
import componentProps from "./props";
import componentEmits from "./emits";
import useImage from "@/composables/useImage";
import useIntersect from "@/composables/useIntersect";
import { MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR } from "@/constants";

const emit = defineEmits(componentEmits);
const props = defineProps(componentProps);

const rootRef = ref(null);
const mainImageRef = ref(null);
const isMainImageRendered = ref(false);
const isFallbackImageRendered = ref(false);

const { isIntersected, watchIntersectionOnce } = useIntersect(rootRef);
const { loadImage, aspectRatio, width } = useImage(mainImageRef);
const isLoading = computed(() => !isMainImageRendered.value);

const paddingHack = computed(() => ({
  paddingBottom: `${aspectRatio.value * 100}%`,
}));

const componentStyle = computed(() => {
  if (props.objectCover || width.value === 0) {
    return;
  }

  return {
    maxWidth: `${width.value}px`,
  };
});

const imageClasses = computed(() => {
  return [
    props.customClass,
    {
      "v-progressive-image-object-cover": props.objectCover,
    },
  ];
});

const onComponentIntersected = () => {
  loadImage()
    .then(() => {
      setTimeout(() => {
        isMainImageRendered.value = true;
        emit(MAIN_IMAGE_LOAD_SUCCESS);
      }, props.delay * 1);
    })
    .catch(() => {
      isMainImageRendered.value = true;
      isFallbackImageRendered.value = true;
      emit(MAIN_IMAGE_LOAD_ERROR, error);
    });
};

onMounted(() => {
  if (props.src) {
    watchIntersectionOnce(onComponentIntersected);
  }
});
</script>

<template>
  <div
    ref="rootRef"
    class="v-progressive-image"
    :class="imageClasses"
    :style="componentStyle"
  >
    <div :style="paddingHack">
      <img
        v-if="isIntersected"
        class="v-progressive-image-main"
        v-show="isMainImageRendered"
        ref="mainImageRef"
        :src="isFallbackImageRendered ? fallbackSrc : src"
        :alt="alt"
      />

      <template v-if="placeholderSrc">
        <svg class="v-progressive-image-svg" v-if="isLoading">
          <filter id="v-progressive-image-filter">
            <feGaussianBlur in="SourceGraphic" :stdDeviation="blur" />
          </filter>
        </svg>
        <transition name="v-progressive-image-fade" appear>
          <img
            v-if="isLoading"
            loading="lazy"
            class="v-progressive-image-placeholder"
            :src="placeholderSrc"
          />
        </transition>
      </template>

      <div v-if="$slots.default" class="v-progressive-image-slot-default">
        <slot :isLoading="isLoading" />
      </div>
    </div>
  </div>
</template>
