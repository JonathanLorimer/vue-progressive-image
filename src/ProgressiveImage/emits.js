import { MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR } from "@/constants";

export default [MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR];
