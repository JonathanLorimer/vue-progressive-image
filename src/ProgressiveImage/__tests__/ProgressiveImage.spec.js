import { mount } from "@vue/test-utils";
import { describe, beforeEach, expect, test, vi } from "vitest";
import ProgressiveImage from "../ProgressiveImage.vue";

describe("ProgressiveImage", () => {
  test("render image", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render image with placeholder", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use custom blur", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        blur: 50,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use alt attribute", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        alt: "image description",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use circle shape", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        circle: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use object cover", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        objectCover: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use object contain", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        objectContain: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use select none", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        selectNone: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render default slot", () => {
    const wrapper = mount(ProgressiveImage, {
      slots: {
        default: "<div>lorem ipsum</div>",
      },
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render default slot props", () => {
    const wrapper = mount(ProgressiveImage, {
      slots: {
        default: `<template #default="params">{{ params }}</template>`,
      },
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("custom aspect ratio", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        aspectRatio: 3.2,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
