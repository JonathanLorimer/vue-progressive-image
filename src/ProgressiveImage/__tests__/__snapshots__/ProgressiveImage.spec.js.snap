// Vitest Snapshot v1

exports[`ProgressiveImage > custom aspect ratio 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 320%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > render default slot 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    <!--v-if-->
    <div
      class="v-progressive-image-slot-default"
    >
      
      <div>
        lorem ipsum
      </div>
      
    </div>
  </div>
</div>
`;

exports[`ProgressiveImage > render default slot props 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    <!--v-if-->
    <div
      class="v-progressive-image-slot-default"
    >
      
      {
  "isLoading": true
}
      
    </div>
  </div>
</div>
`;

exports[`ProgressiveImage > render image 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > render image with placeholder 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use alt attribute 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      alt="image description"
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use circle shape 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 50%; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use custom blur 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(50px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use object contain 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: auto;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main v-progressive-image-object-contain"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use object cover 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: auto;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main v-progressive-image-object-cover"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use select none 1`] = `
<div
  class="v-progressive-image"
  style="border-radius: 0; max-width: 100%;"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <img
      class="v-progressive-image-main v-progressive-image-select-none"
      src="main-image.jpg"
      style="display: none;"
    />
    
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        src="placeholder-image.jpg"
        style="filter: blur(20px);"
      />
    </transition-stub>
    <canvas
      height="1"
      hidden=""
      width="1"
    />
    
    <!--v-if-->
  </div>
</div>
`;
